### ISSUE TITLE
Surprising use of += and function with a side-effect

### ISSUE THREAD

**Issue by mitar** [douglascrockford/JSLint#195]
I just had a surprising experience with the following code:

```javascript
"use strict";
var foo = 0;
function bar() {
    foo += 1;
    return 1;
}
foo += bar();
console.log(foo);
```

I expected for `foo` to be 2 at the end, but it is 1. It would be great if JSLint could detect such cases and warn. That you are doing side-effects to a variable which will not be visible outside because of `+=`.

**Comment by CervEdin** [comment]
This doesn't work as expected since `foo += bar()` equals `foo = foo + bar()` so `foo` is evaluated, before adding the result of the function which invokes `foo += 1`, and adding that result to `foo`.
So any change to `foo` on the LHS would need to happen like this:
```
"use strict";
var foo = 0;
function bar() {
    foo += 1;
    return foo += 1;
}
foo += bar();
console.log(foo);
```
or
```
var foo = 0;
function bar() {
    foo += 1;
    return 1;
}
foo = bar() + foo;
console.log(foo);
```

There's definitely the risk of running into unexpected results but consider

```
var foo = 0;
var baz = 1;
function bar() {
    foo += 1;
    if (foo - baz > 0 ) {
         return 2;
    }
    return 1;
}
foo += bar();
console.log(foo); //1
foo += bar();
console.log(foo); //3
```

This is valid and produces expected results and would throw a warning if implemented like you propose.

**Comment by douglascrockford** [comment]
What are you specifically proposing?
